base/src/main/java/io/spine/code/proto/ref/DirectTypeRef.java

```diff
@@ -0,0 +1,152 @@
+import io.spine.code.proto.PackageName;
+import io.spine.type.TypeName;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.spine.code.proto.ref.BuiltIn.ALL;
+import static io.spine.code.proto.ref.BuiltIn.SELF;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * A direct reference to a proto message type.
+ */
+public final class DirectTypeRef extends AbstractTypeRef {
+
+    private static final long serialVersionUID = 0L;
+
+    private final @Nullable PackageName packageName;
+    private final String simpleTypeName;
+    private final String nestedName;
+
+    /**
+     * Parses the passed value for the subject of direct type reference.
+     *
+     * @return an instance if the passed value is not a wildcard reference or an empty string,
+     *         and empty {@code Optional} otherwise
+     */
+    static Optional<TypeRef> parse(String value) {
+        checkNotNull(value);
+        if (value.equals(SELF.value())) {
+            return Optional.empty();
+        }
+        if (value.contains(ALL.value())) {
+            return Optional.empty();
+        }
+        TypeRef result = new DirectTypeRef(value);
+        return Optional.of(result);
+    }
+
+    private DirectTypeRef(String value) {
+        super(value);
+        String delimiter = PackageName.delimiter();
+        List<String> parts =
+                Splitter.on(delimiter)
+                        .splitToList(value);
+        this.packageName =
+                value.contains(delimiter)
+                ? toPackageName(parts)
+                : null;
+        this.simpleTypeName = parts.get(parts.size() - 1);
+        this.nestedName = toNestedName(parts);
+    }
+
+    /**
+     * Attempts to find a package name in the passed type reference.
+     *
+     * <p>Assumes that package names start from a lowercase letter.
+     *
+     * @return the package name, if found, or {@code null} otherwise
+     */
+    private static @Nullable PackageName toPackageName(List<String> parts) {
+        List<String> packages =
+                parts.stream()
+                     .filter(p -> Character.isLowerCase(p.charAt(0)))
+                     .collect(toList());
+        String result =
+                Joiner.on(PackageName.delimiter())
+                      .join(packages);
+        return PackageName.of(result);
+    }
+
+    /**
+     * Compose a potentially nested type name from the passed parts.
+     *
+     * <p>Assumes that a type name starts from an upper-case letter.
+     */
+    private static String toNestedName(List<String> parts) {
+        List<String> types =
+                parts.stream()
+                     .filter(p -> Character.isUpperCase(p.charAt(0)))
+                     .collect(toList());
+        return Joiner.on(TypeName.NESTED_TYPE_SEPARATOR)
```

Comment:
Why `TypeName.NESTED_TYPE_SEPARATOR` is a public field while `PackageName.delimiter()` is a function?

Reply:
For consistency, because `TypeName` already have another character constant for package separator. Should we decide to unify, both constants should be migrated.